### PR TITLE
drivers: clock_control: avoid calling device_is_ready() in API calls

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Change the clock control driver API to not call device_is_ready() in every API call. It is up to the caller to ensure that the clock control device is ready before calling a clock control API function.
    
Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
I did not test all of these changes, so reviews and additional testing are most welcome.

File: include/zephyr/drivers/clock_control.h [hunk 1]
```diff
@@ -43,7 +43,6 @@ enum clock_control_status {
 	CLOCK_CONTROL_STATUS_STARTING,
 	CLOCK_CONTROL_STATUS_OFF,
 	CLOCK_CONTROL_STATUS_ON,
-	CLOCK_CONTROL_STATUS_UNAVAILABLE,
```
Comment (henrikbrixandersen) [inline include/zephyr/drivers/clock_control.h:46]: I have removed this as the only in-tree usage was to indicate a non-ready clock control device. We can leave it in if people suspect out-of-tree users, but I believe `CLOCK_CONTROL_STATUS_UNKNOWN` would be good enough for such usage.

**Review by cfriedt — DISMISSED** [review]

**Review by gmarull — DISMISSED** [review]
thanks for this cleanup

**Review by martinjaeger — DISMISSED** [review]
Thanks!

File: drivers/i2c/i2c_lpc11u6x.c [hunk 2]
```diff
@@ -314,6 +314,11 @@ static int lpc11u6x_i2c_init(const struct device *dev)
 		return err;
 	}
 
+	if (!device_is_ready(cfg->clock_dev)) {
+		LOG_ERR("clock control device not ready");
```
Comment (anangl) [inline drivers/i2c/i2c_lpc11u6x.c:318]: Logging does not seem to be enabled for this file. Can this be actually compiled successfully?

Comment (henrikbrixandersen) [reply drivers/i2c/i2c_lpc11u6x.c]: You're right, thanks. Fixed.

**Review by anangl — APPROVED** [review]

**Comment by henrikbrixandersen** [comment]
@cfriedt, @martinjaeger, @gmarull: Please revisit.

**Comment by cfriedt** [comment]
Now that I see all of the other places where callers (device drivers mainly) need to verify a clock is already initialized before continuing, I'm starting to wonder if this is more of a broader issue that should be solved in a different way.

FWIR, part of the ordering that happens in the post-link phase (kind of an ELF hack) is similar to a topological sort / dependency resolution.

It's clear that the ordering works well for nested nodes in DT, but it's possible that the clock reference is just a phandle and that ordering might not be as easy in that case.

Of course, we have multiple device init priorities, so maybe the better solution is to have the earliest device init priority be reserved exclusively for clocks.

In that case, we could omit the `device_is_ready()` calls (or potentially surround them in an `__ASSERT()`.

**Comment by henrikbrixandersen** [comment]
> In that case, we could omit the `device_is_ready()` calls (or potentially surround them in an `__ASSERT()`).

The driver initialization function for a clock controller driver could still fail, causing the device to not be ready for use.

**Comment by galak** [comment]
> Now that I see all of the other places where callers (device drivers mainly) need to verify a clock is already initialized before continuing, I'm starting to wonder if this is more of a broader issue that should be solved in a different way.
> 
> FWIR, part of the ordering that happens in the post-link phase (kind of an ELF hack) is similar to a topological sort / dependency resolution.
> 
> It's clear that the ordering works well for nested nodes in DT, but it's possible that the clock reference is just a phandle and that ordering might not be as easy in that case.
> 
> Of course, we have multiple device init priorities, so maybe the better solution is to have the earliest device init priority be reserved exclusively for clocks.
> 
> In that case, we could omit the `device_is_ready()` calls (or potentially surround them in an `__ASSERT()`).

I felt like @JordanYates was possibly working on something in this area with the devhandles work.

**Review by gmarull — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]

**Comment by carlescufi** [comment]
API meeting:

- Need @nordic-krch approval
- Additional release note warning users to perform the check themselves in a follow-up PR
- No additional objections

**Comment by cfriedt** [comment]
> > In that case, we could omit the `device_is_ready()` calls (or potentially surround them in an `__ASSERT()`).
> 
> The driver initialization function for a clock controller driver could still fail, causing the device to not be ready for use.

Hmm.. yeah, I guess it might not always be considered fatal.

**Comment by JordanYates** [comment]
> > Now that I see all of the other places where callers (device drivers mainly) need to verify a clock is already initialized before continuing, I'm starting to wonder if this is more of a broader issue that should be solved in a different way.
> > FWIR, part of the ordering that happens in the post-link phase (kind of an ELF hack) is similar to a topological sort / dependency resolution.
> > It's clear that the ordering works well for nested nodes in DT, but it's possible that the clock reference is just a phandle and that ordering might not be as easy in that case.
> > Of course, we have multiple device init priorities, so maybe the better solution is to have the earliest device init priority be reserved exclusively for clocks.
> > In that case, we could omit the `device_is_ready()` calls (or potentially surround them in an `__ASSERT()`).
> 
> I felt like @JordanYates was possibly working on something in this area with the devhandles work.

Broadly yes. The idea being to eventually transition the manual init level specification to a dependency graph traversal.

**Pull request merged**